distort-fisheye.cpp

```diff
@@ -0,0 +1,67 @@
+#include <opencv2/opencv.hpp>
+#include <string>
+#include <vector>
+
+using namespace cv;
+using namespace std;
+
+int main(int argc, char* argv[])
+{
+    cout << "2. Distort: input camera params and one undistorted image, then get the distorted image." << endl;
+
+    string fn_image = "undistort.png";
+    string fn_param = "fisheye-params.json";
+    if (argc > 1)
+        fn_image = argv[1];
+    if (argc > 2)
+        fn_param = argv[2];
+    Mat image_ud = imread(fn_image, IMREAD_GRAYSCALE);
+    if (image_ud.empty()) {
+        cout << "cannot load image: " << fn_image << endl;
+        return 0;
+    }
+    Mat cam_mtx, dis_cef;
+    Size image_size;
+    auto loadParams = [&]() {
+        FileStorage fs;
+        if (!fs.open(fn_param, FileStorage::READ))
+            return false;
+        fs["cam_mtx"] >> cam_mtx;
+        fs["dis_cef"] >> dis_cef;
+        fs["image_size"] >> image_size;
+        return !cam_mtx.empty() && !dis_cef.empty() && image_size.area() > 0;
+    };
+    if (!loadParams()) {
+        cout << "cannot load camera params from " << fn_param << endl;
+        return 0;
+    }
+
+    Mat R;
+    Mat distorted_cam_mtx = cam_mtx.clone();
+
+    // add distortion
+    Mat map_x = Mat(image_size, CV_32FC1);
+    Mat map_y = Mat(image_size, CV_32FC1);
+    vector<Point2f> distorted_output_pts, undistorted_input_pts;
+    for (int y = 0; y < image_size.height; ++y)
+        for (int x = 0; x < image_size.width; ++x)
+            undistorted_input_pts.emplace_back(x, y);
+    fisheye::undistortPoints(undistorted_input_pts, distorted_output_pts, cam_mtx, dis_cef, R, distorted_cam_mtx);
```

Comment:
In fact, the output pts of `undistortPoints` should be called **undistorted_pts**. It represents the coordinate-map from the *distorted image* to the *undistorted image*.